_posts/2020-08-17-creating-your-ci-pipeline-with-concource.md

```diff
@@ -0,0 +1,217 @@
+in your variables. [Concourse supports many different credential managers out-of-the-box](https://concourse-ci.org/creds.html).
+By using external manager you only need to provide api key when you are creating your pipeline.
+In general, when you are going to set up new CI-system, you should always do a very careful evaluation from the
+security point of view. However, this is out of the scope of this blog post.
+
+
+Right, now we have some input. Goal in this example is to get some input from git,
+make some *things* and then push the result (or output) into specific repository in Docker Hub.
+So, next we define an output. In general, output can be whatever is needed,
+ie. rsync your project artifacts to some server, push something to git and so on.
+
+```yaml
+- name: concource-example-registry
+  type: registry-image-resource
+  source:
+    repository: ((docker-source))
+    tag: ((docker-tag))
+    username: ((docker-username))
+    password: ((docker-password))
+```
+
+So, now we have external input and output. Next phase is to define a job which uses those. Let's use
+our imagination and call our job *build-and-publish*. You could define as many jobs as you wish,
+but in this case one is enough.
+
+```yaml
+jobs:
+- name: build-and-publish
+```
+
+Every job has some series of steps which is called a *plan*.
+
+```yaml
+plan:
+    - get: concource-example
+      trigger: ((trigger-build))
+    - task: prepare-build
+      file: concource-example/concource/pipelines/concource-example/concource-example-prepare-build.yml
+      vars:
+        docker-file: ((docker-file))
+        docker-tag: ((docker-tag))
+    - task: build
+      file: concource-example/concource/pipelines/concource-example/concource-example-build.yml
+      privileged: true
+      params:
+        DOCKERFILE: prebuild-output/((docker-file))
+    - put: concource-example-registry
+      params: {
+        image: image/image.tar
+      }
+```
+
+So, while this definition is very self-explanatory let's walk it through anyway. First we fetch our input
+from Git. Then we do some preliminary *things* before building our complex Docker image. Finally, we
+put image into the Docker Hub. One point to note is keyword *trigger* in our first *get*. This boolean value
+controls if the job should be triggered automatically when resource state changes
+(Ie. you push something into your repository).
+
+I have divided my more complex tasks into separate files, so it's easier to see top-level flow from main pipeline file.
+When creating separate task files, it's important to understand, that only main pipeline definition will be stored in
+system memory. Everything else, including separate task definition files, will be fetched from some external resource
+(Ie. Git in our case). This means that when you are about to update your tasks, you have to push those into external
+resource to changes take an effect. 
```

Comment:
a top-level
from the main
only the main
into an external resource/into external resources